### PR TITLE
win32: make context menu items accessible from the window menu

### PR DESCRIPTION
**Pull request by na-na-hi**
This makes the context menu items accessible from the window menu, which can be opened by either right-clicking on the title bar or left-clicking on the mpv icon on the title bar.

### CODE REVIEW

**Comment by kasper93** [comment]
I frankly don't like appending user items to system menu. If you click in client area you get the client menu, else you get system one. What would be the use case to merge them?

**Comment by kasper93** [comment]
> It does though? At least for clsid2's fork.

Forget about it, but my point was that MPC-HC has huge UI elements that are not connected to play/pause. Also having play/pause on left click is very often conflicting with having fullscreen on left double click. That's another topic.

> conhost.exe does.

Yes, but it has different menu there. This was my initial concern depending on the user menu the system menu might get big, same as the other.

> Also since "Exit" is usually the last item of a context menu (which is the case for MPC-HC), putting it just before "Close" seems a bit weird. Now we have 2 consecutive items doing the same thing.

That's why merging those menus is kinda tricky, because not all elements make sense to be in one or the other. For example the same argument can be made for general case when someone makes menu with maximize/minimize/close items, now you get them duplicated.

I don't see this as big usability improvement, but at the same time, it is ok to have it if you feel like it.

**Comment by tsl0922** [comment]
> I frankly don't like appending user items to system menu.

So do I.

> how about making the whole context menu a submenu in the system menu?

Maybe add an option that allow the whole system menu to be replaced with context menu is more flexible? since you can already change the structrure of the context menu. The new option may have 3 values: append / replace / no.

**Comment by tsl0922** [comment]
Well, I still think we don't need to put it in a submenu, you can already change the structure of the context menu (make all of it's items under a top level `Context Menu` item) with scripts. This also allow users to change the `Context Menu` string to a localized one.

**Comment by tsl0922** [comment]
> The problem here is that the context menus for both cases have the same data.

Sorry, I forget it 😄 . My original concern was the `Context Menu` string is hardcoded, which make it impossible to localize it.

**Comment by tsl0922** [comment]
I'm OK with both solutions.

**Comment by kasper93** [comment]
> Make the context menu a submenu in the system menu, and add another property menu-title to set the title (which replaces the "Context Menu" string here).

title could be `mpv` which is not suppose to be translated. This solves the problem of localization.

> Append all context menu items to the window menu, and add --win32-window-menu option which allows the user to disable the modification of the system menu. The user is responsible to make sure that the contents of the menu make sense for either the popup or the system menu, depending on preferences.

Option doesn't harm, but I don't think it really solves the problem, just allows users to disable it if they don't like it.


From my side I'm fine with always appending context menu to system menu or nest it in `mpv` item for simplicity. Can be made configurable if you want it to be.

**Comment by kasper93** [comment]
Menu doesn't work for me after those changes.

**Comment by kasper93** [comment]
> > Menu doesn't work for me after those changes.
> 
> Can you elaborate what doesn't work? The latest CI build works for me on Windows 10 and 11. I used the `test.lua` in #13700 for testing.

This indeed works, but I'm using https://github.com/tsl0922/mpv-menu-plugin/blob/21825a1709d9168731f4d79983fd46c652e6f67c/src/lua/dyn_menu.lua to test and it just doesn't do anything anymore. Didn't debug it in anyway, so I don't know at this point why.

**Comment by kasper93** [comment]
Maybe add a separator before `mpv` submenu?

**Comment by kasper93** [comment]
If there is no menu defined, `mpv` should not be added imo.

File: video/out/win32/menu.c [hunk 1]
```diff
@@ -178,9 +180,10 @@ static void build_menu(void *talloc_ctx, HMENU hmenu, struct mpv_node *node)
...
 }
 
-struct menu_ctx *mp_win32_menu_init(void)
+struct menu_ctx *mp_win32_menu_init(HWND hwnd)
 {
     struct menu_ctx *ctx = talloc_ptrtype(NULL, ctx);
+    ctx->hwnd = hwnd;
     ctx->menu = CreatePopupMenu();
     ctx->ta_data = talloc_new(ctx);
     return ctx;
```
Comment (kasper93) [inline video/out/win32/menu.c:189]: ctx->update is not initialized.

Comment (kasper93) [reply video/out/win32/menu.c:189]: Also, `set menu-data null` should probably remove the already added submenu in system menu.

File: video/out/win32/menu.c [hunk 2]
```diff
@@ -218,6 +222,17 @@ void mp_win32_menu_update(struct menu_ctx *ctx, struct mpv_node *data)
     talloc_free_children(ctx->ta_data);
 
     build_menu(ctx->ta_data, ctx->menu, data);
+    if (!ctx->updated) {
```
Comment (tsl0922) [inline video/out/win32/menu.c:225]: What about making a copy of ctx->menu for system menu here to make it easier:

1. always reset system menu
2. check menu item count, if >0 then append submenu.

**Review by kasper93 — APPROVED** [review]

**Pull request merged**